### ISSUE TITLE
Replace Authlogic and Implement Invites

### ISSUE THREAD

**Issue by sudara** [sudara/alonetone#874]
This is the new #838 

As public signups are closed, and will be closed for the foreseeable future, we need a way to get accounts to musicians. The goal is to build a very simple invite system. 

While we are at it, it may make sense to pull Authlogic and replace with some straightforward custom authentication. 

Requirements for invites:

* [ ] People can publicly add themselves to an invite waiting list as described in #871, which will be designed this weekend. We are collecting a bit of additional information there, so it probably make sense to have an invites table and these invites are pending. The additional information is actually not really needed however, it's essentially for screening purposes (we only want to allow artists in, not podcasters/preachers/bloggers).
* [ ] When an existing users' account is older than 2 weeks, has at least 2 tracks and has at least 2 login_counts, then we want to offer them the ability to invite users. We want to track who invites whom, but I don't believe we need limits on number of invites at this point in time. Perhaps in the future. When public signups were still open, we wanted to move forward with #652, which was to whitelist users as being "active musicians," which is sort of similar but different to this ability.  
* [ ] All moderators should be able to send an invite from the admin area
* [ ] In the Admin area, we need ability to see all pending invites and mark them as approved, which will then send an invite email where they can choose a password. 
* [ ] In the Admin area, we want to see all invites that were sent, see which ones were claimed and be able to re-send an email.
* [ ] A moderator should have the ability to create an invite URL, for sharing on social media and custom emailing. This is a link that allows account creation and is not tied to an email. 
* [ ] We may, at some future point in time, want to show who invited whom. 

If Authlogic is pulled, it will fix #833 and #822, which will be nice. Here are the minimum requirements for replacing:

* [ ] Users can login with their 'login' and password. 
* [ ] Users can still login with their current password. Using `Authlogic::CryptoProviders::SCrypt` right now.
* [ ] Email verification is a thing. Clicking the verification link in email automatically logs them in. Someone should be able to verify an account even if they are already logged in. See #845
*. [ ] Forgot Your Password? sends an email with a new verification link that logs the user in and takes them to the password reset form. 
* [ ] Forgot Your Login? This does happen and generates support requests, but I'm not 100% sure we need to build for it. People usually search for themselves or url hack to remember.  
* [ ] We keep `current_user` as the helper...
* [ ] Continue to properly maintain login_count, current_login_at, current_login_ip, last_request_at. I'm not 100% sure when last_request_at updates currently, I think on most requests? We could also do "last_session_at" to cut down on excessive updates. It's useful and nice to see who is on the site, and we display this in the UI.

Loose ends:

* [ ] We need a migration to backfill whether or not a user is whitelisted as being able to invite others. 
* [ ] lib/moderator_constraint
* [ ] Ensure thredded is still happy

**Comment by sudara** [comment]
@Manfred Thought this might be "meaty" enough for you. It also would be extremely helpful, as right now signups are closed and I have to manually create accounts and reset perishable tokens, etc. After we get this up and running, I plan to start inviting people into alonetone again.

Samo and I are working on the waiting list/application UI tomorrow. Everything else UI-wise should be pretty plain forms. Ben is happy to jump into an open branch to manage and adjust styles if you tag him or I.

**Comment by Manfred** [comment]
@sudara Do you want to switch to Argon2 now we're touching the code anyway or do you feel sticking with SCrypt is safe enough for now?

**Comment by sudara** [comment]
@manfred Hmmmm.... I'm the kind of person who tends to travel with the masses vs. vanguard on calls like this. But if you are a fan and it's not a difficult/verbose dependency to add, go ahead!

**Comment by Manfred** [comment]
I'm confused by how the word invite is used in this issue. It seems like at some points it refers to the implementation (eg. a database record) and sometimes it refers to the act of inviting someone.

So from a purely conceptual standpoint: are users always going to fill out an application form, or will existing users invite new users, or both?

**Comment by sudara** [comment]
Hey there. Sorry about that. Yes, I'm using "invite" as a noun that means possessing the ability to enter as well as "invite" as a verb meaning to create a user account and send off the invite. 

Part of the confusion I think is that the "application form" creates a request for an invite which is a subtle difference — but the actual invite still has to be manually triggered by a mod.

> are users always going to fill out an application form, or will existing users invite new users, or both?

There are 4 ways for a user to get on the platform:

1) They submit an application, which goes into a moderation queue and is exposed in `/admin/users`, at which point a moderator can "approve and send invite". This ensures everyone has a chance to get on the platform.

2) A moderator can directly send an invite to an email address (and we remember who invited them).

3) A moderator can generate an one time use url which allows it to be given away. This is useful for posting to groups/forums/reddit or to places where we don't have the recipient's email, which is an important part of our strategy for growing again. We'd need to perform email validation in this case.

4) "Whitelisted" users who have passed an engagement threshold can invite users via email, just like moderators can, and we remember who invited whom.

**Comment by sudara** [comment]
It looks like we are on the same page! Thanks for booting this into your brain.

A couple thoughts:

-1. Artist is the best name! 
0. We could keep `Profile` around or shuffle it back into an `Artist` model. `user_agent` is the only thing that should go hang out with `User` 
1. My vote is to go out of our way to pick something less generic than `UserApplication` maybe like `SignupRequest` or `RegistrationRequest`. I don't have any great ideas, but `Application` is a bit confusing in a backend context. This will have a few more attributes on it for the form in #871 
2. `login` is currently both the user identifier and the artist slug/name, both of which need to be unique. I'm not sure it makes sense to keep `login` as a way to get into the app, as they are all artist names. It's ok to start with that as the transition while we are 1:1 user:artist. But it might make more sense that we have people log in with their email/pass and migrate `login` to be the artist `slug`.
3. So.... we probably need a `slug` and a `name` on `Artist` to match up with `login` and `display_name` (which gets populated with login on account creation, able to be changed later).
4. I'm less clear on the purpose of `Role` (assuming it's so we can do many to many down the line) and `RoleInvitation` (assuming it's so you could invite people to co-manage your artist) but I think we have the option to consider those out of scope and go with one user -> many artists for now. I like the idea of eventually allowing users to co-manage an artist, but it's the cherry on top of the cherry on top of the cherry. It will be *such* a huge step up to be able to switch between multiple artist identities. It's far less common to co-manage identities, though it's regular and does happen. 
5. I'm fine with how we were storing sessions before, in a cookie. 
6. No plans or thoughts about multi-factor or auth via some big corp for identity. I think we can keep scope tight here.

**Comment by sudara** [comment]
After thinking about this a bit, maybe we should reduce scope for now to just do the invites and wait on the multi-artist accounts. 

Part of what I'm thinking is that it might be a lot of work to get every relation, including soft deletion to be pointing towards artists. I'm not sure what a good strategy is there, but it seems high effort. Seeing as the priority is we want to get people back onto the platform, maybe we should wait on multi-artist accountn and just know it's coming?

**Comment by Manfred** [comment]
The plan was never to change this all in one go. I just wanted to draft up an idea of the future so there is some context when we need to make decisions.

**Comment by sudara** [comment]
Yeah, was mainly expressing I’d prefer to get invites going more quickly
vs. delay it due to refactoring for a “would be nice down the line”
feature.

But it’s good to talk through: with your help, I feel that more
opportunities are possible than in the past. I could never imagine getting
around to prioritizing multi-artist support pre-Manfred!

**Comment by sudara** [comment]
Update:

* I still slightly prefer replacing Authlogic in the medium term, due to issues like https://github.com/binarylogic/authlogic/issues/693 — things like passwords are completely managed by authlogic, we keep having to discover and work around issues on Authlogic's terms. We can play it by ear and see if it impacts invites/multi-artist work as well.

* Account Requests were merged in #945 with some followup work described in #954

That leaves:

* User ability (after passing usage threshold) and moderator ability to invite users 
* More than one artist per account
